working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,752 @@
+clause, as described above.
+
+In a show or hide clause, it is possible that an
+`<extensionShowHideElement>` is an identifier that is the basename of a
+member of the interface of the on-type, and it is also the name of a type
+in scope. In this case, the name shall refer to the member.
+
+*A conflict is unlikely because type names in general are capitalized, and
+member names start with a lower-case letter. Some type names start with a
+lower-case letter, too (e.g., `int` and `dynamic`), but those names do not
+occur frequently as member names. Should a conflict arise anyway, a
+work-around is to import the shadowed type `T` with a prefix `p` and put
+`p.T` in the show or hide clause.*
+
+A compile-time error occurs if a hide or show clause contains an identifier
+which is not the basename of an instance member of the on-type. A
+compile-time error occurs if a hide or show clause contains a type which is
+not among the types that are implemented by the on-type of the extension.
+
+A type in a hide or show clause may be raw (*that is, an identifier or
+qualified identifier denoting a generic type, but no actual type
+arguments*). In this case the omitted type arguments are determined by the
+corresponding superinterface of the on-type.
+
+*For example:*
+
+```dart
+extension type MyInt on int show num, isEven hide floor {
+  int get twice => 2 * this;
+}
+
+void main() {
+  MyInt m = 42;
+  m.twice; // OK, in the extension type.
+  m.isEven; // OK, a shown instance member.
+  m.ceil(); // OK, a shown instance member.
+  m.toString(); // OK, an `Object?` member.
+  m.floor(); // Error, hidden.
+}
+```
+
+
+### Implementing superinterfaces
+
+This section specifies the effect of having an `<interfaces>` part in an
+extension declaration.
+
+Let `E` be an extension declaration where `<interfaces>?` is of the form
+`implements T1, .. Tm`. We say that `E` has `T1` .. `Tm` as its direct
+superinterfaces.
+
+A compile-time error occurs if a direct superinterface does not denote a
+class, or if it denotes a class which cannot be a superinterface of a
+class.
+
+*For instance, `implements int` is an error.*
+
+For each member `m` named `n` in each direct superinterface of `E`, an
+error occurs unless `E` declares a member `m1` named `n` which is a correct
+override of `m`, or the show/hide part of `E` enables an instance member of
+the on-type which is a correct override of `m`.
+
+No subtype relationship exists between `E` and `T1, .. Tm`.
+
+*This means that when an extension type implements a set of interfaces, it
+is enforced that all the specified members are available, and that they
+have a signature which is compatible with the ones in `T1, .. Tm`, but
+there is no assignability from an expression of type `E` to a variable
+whose declared type is `Tj` for some `j` in 1..m. For that, it is necessary
+to use `box`, as described below.*
+
+
+### Invariant enforcement through introduction: Protected extension types
```

Comment:
As mentioned later, drop the "invariant" word, it's a distraction assuming a specific use-case. Just "restrict extension type values".